fsthttp/request.go

```diff
@@ -288,33 +290,54 @@ func (req *Request) Send(ctx context.Context, backend string) (*Response, error)
 		}
 	}
 
-	abiPending, abiReqBody, err := req.sendABIRequestAsync(backend)
-	if err != nil {
-		return nil, err
+	if req.sent {
+		return nil, fmt.Errorf("request already sent")
 	}
+	req.sent = true
 
 	var (
-		errc         = make(chan error, 3)
-		bodyExists   = req.Body != nil
-		_, bodyIsABI = req.Body.(*fastly.HTTPBody)
-		shouldCopy   = bodyExists && !bodyIsABI
+		errc      = make(chan error, 3)
+		streaming = true
 	)
 
-	if shouldCopy {
-		go func() {
-			_, copyErr := io.Copy(abiReqBody, req.Body)
-			errc <- maybeWrap(copyErr, "copy body")
-			errc <- maybeWrap(req.Body.Close(), "close user body")
-			if copyErr == nil {
-				// If there was an error copying the body, we *don't* want to Close() the abi req.
-				// This tells the wasm server that the body is incomplete so it knows not to
-				// terminate the sent chunked body with a valid final chunk.
-				// Thus, only Close if copyErr == nil.
-				errc <- maybeWrap(abiReqBody.Close(), "close request body")
-			}
-		}()
+	// When the request's ManualFramingMode is false, SendAsyncStreaming
+	// streams the request body to the backend using "Transfer-Encoding:
+	// chunked".  SendAsync buffers the entire body and sends it with a
+	// "Content-Length" header.
+	//
+	// For requests without a body we want to avoid unnecessary chunked
+	// encoding, and have observed servers that error when seeing it in
+	// certain contexts.
+	//
+	// For requests where the body is an io.Reader implementer where the
+	// size is known in advance, we want to send that along with
+	// Content-Length as well.  Those types are *bytes.Buffer,
+	// *bytes.Reader, and *strings.Reader.
+	//
+	// For all other requests, we stream with chunked encoding.
+	switch underlyingReaderFrom(req.Body).(type) {
+	case nil, *bytes.Buffer, *bytes.Reader, *strings.Reader:
```

Comment:
maybe check whether it implements the Lenner interface (which I just made up)?

Reply:
I thought about this, and I checked the stdlib for all `io.Reader`s that implemented that interface, and these are the only 3 exported ones that came up.  (There are a handful of internal ones, particular in the `net/http` package.)

Ultimately I decided not to do this because I felt that having `Len() int` wasn't a strong enough contract for an in-memory buffer.  You could imagine an `io.Reader` that streams data from the network, but the size is known ahead of time because it was sent by the client in something akin to `Content-Length` and it was exposed this way.  We'd want to stream the data in that case.